### PR TITLE
[#15] [Documentation] Create project infrastructure diagram

### LINKED ISSUE #15: [Documentation] Create project infrastructure diagram
## Why

Creating an infrastructure diagram can be beneficial for a number of reasons. It can help you understand the relationships and dependencies between different components of your infrastructure, which can be useful when troubleshooting issues or making changes to the system. It can also be a helpful tool for onboarding new team members, as it provides a visual overview of the infrastructure.

Additionally, an infrastructure diagram can be useful for documenting and communicating the architecture of your system to stakeholders, such as clients or upper management. It can also be useful for identifying potential areas for improvement or optimization.

## Acceptance Criteria

- The infrastructure diagram should accurately represent the future infrastructure of the project, including all relevant components and their relationships.
- The diagram should be easy to understand and clearly labeled.
- The diagram should include details such as network topology, server specifications and third party services

### PR DESCRIPTION
**Pull request by Nihisil**
Closes #15

## What happened 👀

A full infrastructure diagram was created prior to the implementation phase in order to gain a thorough understanding of the planned infrastructure.

## Insight 📝

[Diagrams](https://diagrams.mingrammer.com) Python library was used to generate diagram. It will help to track changes in diagram since it will be placed to the source control system.

## Proof Of Work 📹

Diagram:

image

### CODE REVIEW

File: .github/wiki/Infrastructure-Diagram.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+The infrastructure is managed using Terraform, and the demo app can be found here https://github.com/Nihisil/nimble-devops-ic-app.
```
Comment (longnd) [inline .github/wiki/Infrastructure-Diagram.md:1]: This wiki page will not be published automatically without a workflow, e.g. https://github.com/nimblehq/publish-github-wiki-action
I assume you will do that in another story, given that the focus of this PR is to create the diagram only, is that correct?

File: .github/wiki/Infrastructure-Diagram.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+The infrastructure is managed using Terraform, and the demo app can be found here https://github.com/Nihisil/nimble-devops-ic-app. 
+
+Here is a diagram illustrating the infrastructure:
+
+![infrastructure diagram](bin/infrastructure_diagram.png)
```
Comment (longnd) [inline .github/wiki/Infrastructure-Diagram.md:5]: just a suggestion, as the library used to generate the diagram supports multiple formats to export
https://diagrams.mingrammer.com/docs/guides/diagram#options

it's worth exporting the diagram as SVG as it scales better, e.g. the shapes do not look pixelated :D

File: .github/wiki/bin/create_diagram.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+from diagrams.aws.network import ALB
+from diagrams.aws.storage import S3
+from diagrams.aws.security import SecretsManager
+from diagrams.aws.compute import EC2Instance
+from diagrams.aws.management import Cloudwatch
+from diagrams.generic.network import Subnet
+from diagrams.aws.compute import ECR
+
+
+def main():
```
Comment (longnd) [inline .github/wiki/bin/create_diagram.py:18]: are we missing the main part - where the app is hosted, e.g. the ECS Fargate cluster?
I assume it is abstracted by the Autoscaling group?

File: .github/wiki/bin/create_diagram.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    with Diagram("Infrastructure Diagram", show=False):
+
+        SecretsManager("Secrets Manager")
+        ECR("Container registry")
```
Comment (longnd) [inline .github/wiki/bin/create_diagram.py]: Also relates to the above question - the diagram does not show the relationship between the ECS & the auto-scaling group (how they link together), e.g. if you use Fargate, it needs the ECR to store the container image of the app, right? so it's better to show that link.

File: .github/wiki/Infrastructure-Diagram.md (hunk 2 above)
Comment (Nihisil) [reply .github/wiki/Infrastructure-Diagram.md:5]: Actually, I tried to do so, but it generates a diagram without icons. I don't know if something is missing for me or if it is an issue of the library, so I decided to leave it as png for the first version.

image

File: .github/wiki/Infrastructure-Diagram.md (hunk 1 above)
Comment (Nihisil) [reply .github/wiki/Infrastructure-Diagram.md:1]: Yep, wiki workflow will be set up during this ticket: https://github.com/Nihisil/nimble-devops-ic-infrastructure/issues/1

File: .github/wiki/bin/create_diagram.py (hunk 3 above)
Comment (Nihisil) [reply .github/wiki/bin/create_diagram.py:18]: I made the Fargate cluster explicit to be shown here: https://github.com/Nihisil/nimble-devops-ic-infrastructure/pull/16/commits/3668e2c71bb56b1cd526301d6afad6718ebba4bb

Previously it was abstracted by the autoscaling group, as you guessed.

File: .github/wiki/bin/create_diagram.py (hunk 4 above)
Comment (Nihisil) [reply .github/wiki/bin/create_diagram.py]: The link between ECR and Fargate groups was added here: https://github.com/Nihisil/nimble-devops-ic-infrastructure/pull/16/commits/3668e2c71bb56b1cd526301d6afad6718ebba4bb

**Review by longnd — APPROVED** [review]

**Review by andyduong1920 — APPROVED** [review]

File: .github/wiki/Infrastructure-Diagram.md (hunk 2 above)
Comment (longnd) [reply .github/wiki/Infrastructure-Diagram.md:5]: for future reference, this is a know issue that has not been fixed
* https://github.com/mingrammer/diagrams/issues/764
* https://github.com/mingrammer/diagrams/issues/8

**Pull request merged**